Unit test fails under macOS

# Description
```
[ERROR] Failures: 
[ERROR]   TestOriginLinkFormatter.testFormatAbsoluteUnixPathWithSpecialCharacters:104->assertPathAndLineRenderedToSpan:59 
Expected: "<span class=\"origin\"><a href=\"file:///fo%20o/b%C3%A4r/baz\">/fo o/bär/baz</a>:12345678</span>"
     but: was "<span class=\"origin\"><a href=\"file:///fo%20o/ba%CC%88r/baz\">/fo o/bär/baz</a>:12345678</span>"
```

## Steps to Reproduce
`mvn clean test` under macOS

--- a/doc/CHANGELOG.md
+++ b/doc/CHANGELOG.md
@@ -25,6 +25,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Upgrade test and build dependencies [PR #330](https://github.com/itsallcode/openfasttrace/pull/330)
 - Removed license headers from sources and license-maven-plugin [PR #331](https://github.com/itsallcode/openfasttrace/pull/331)
 
+### Bugfixes
+
+- Fixed running unit tests under macOS [#308](https://github.com/itsallcode/openfasttrace/issues/308) / (PR #339)[https://github.com/itsallcode/openfasttrace/pull/339]
+  - Also fixed Jacoco code coverage and integration tests
+
 ## [3.5.0] - 2022-03-17
 
 ### Updated

--- a/doc/CHANGELOG.md
+++ b/doc/CHANGELOG.md
@@ -27,7 +27,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Bugfixes
 
-- Fixed running unit tests under macOS [#308](https://github.com/itsallcode/openfasttrace/issues/308) / (PR #339)[https://github.com/itsallcode/openfasttrace/pull/339]
+- Fixed running unit tests under macOS [#308](https://github.com/itsallcode/openfasttrace/issues/308) / [PR #339](https://github.com/itsallcode/openfasttrace/pull/339)
   - Also fixed Jacoco code coverage and integration tests
 
 ## [3.5.0] - 2022-03-17

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -14,6 +14,7 @@
         <maven.surefire.version>3.0.0-M7</maven.surefire.version>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+        <jacoco.version>0.8.8</jacoco.version>
     </properties>
     <licenses>
         <license>
@@ -338,8 +339,33 @@
                 </configuration>
             </plugin>
             <plugin>
-                <groupId>org.pitest</groupId>
-                <artifactId>pitest-maven</artifactId>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <version>${maven.surefire.version}</version>
+                <configuration>
+                    <systemPropertyVariables>
+                        <java.util.logging.config.file>${basedir}/src/test/resources/logging.properties</java.util.logging.config.file>
+                    </systemPropertyVariables>
+                    <useModulePath>true</useModulePath>
+                    <argLine>
+                        @{argLine}
+                        --illegal-access=deny
+                        <!-- Required to fix failing EqualsVerifier test:
+                            TestTracedLink.testEqualsAndHashContract:22 EqualsVerifier found a problem in class org.itsallcode.openfasttrace.api.core.TracedLink.
+                            -> Unable to make field private final org.itsallcode.openfasttrace.api.core.LinkedSpecificationItem org.itsallcode.openfasttrace.api.core.TracedLink.otherLinkEnd accessible:
+                            module org.itsallcode.openfasttrace.api does not "opens org.itsallcode.openfasttrace.api.core" to unnamed module @48ae9b55 -->
+                        --add-opens org.itsallcode.openfasttrace.api/org.itsallcode.openfasttrace.api.core=ALL-UNNAMED
+                    </argLine>
+                </configuration>
+                <executions>
+                    <execution>
+                        <id>verify</id>
+                        <goals>
+                            <goal>integration-test</goal>
+                            <goal>verify</goal>
+                        </goals>
+                    </execution>
+                </executions>
             </plugin>
             <plugin>
                 <groupId>org.codehaus.mojo</groupId>

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -250,6 +250,7 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.9.0</version>
                 <configuration>
                     <source>${java.version}</source>
                     <target>${java.version}</target>
@@ -311,6 +312,7 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
+                <version>${maven.surefire.version}</version>
                 <configuration>
                     <systemPropertyVariables>
                         <java.util.logging.config.file>${basedir}/src/test/resources/logging.properties</java.util.logging.config.file>
@@ -337,6 +339,7 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-enforcer-plugin</artifactId>
+                <version>3.0.0</version>
                 <executions>
                     <execution>
                         <id>enforce-maven</id>
@@ -383,18 +386,8 @@
                 </plugin>
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
-                    <artifactId>maven-compiler-plugin</artifactId>
-                    <version>3.9.0</version>
-                </plugin>
-                <plugin>
-                    <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-jar-plugin</artifactId>
                     <version>3.2.2</version>
-                </plugin>
-                <plugin>
-                    <groupId>org.apache.maven.plugins</groupId>
-                    <artifactId>maven-surefire-plugin</artifactId>
-                    <version>${maven.surefire.version}</version>
                 </plugin>
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
@@ -405,11 +398,6 @@
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-javadoc-plugin</artifactId>
                     <version>3.3.1</version>
-                </plugin>
-                <plugin>
-                    <groupId>org.jacoco</groupId>
-                    <artifactId>jacoco-maven-plugin</artifactId>
-                    <version>0.8.7</version>
                 </plugin>
                 <plugin>
                     <groupId>org.codehaus.mojo</groupId>
@@ -441,21 +429,11 @@
                     <artifactId>maven-deploy-plugin</artifactId>
                     <version>3.0.0</version>
                 </plugin>
-                <plugin>
-                    <groupId>org.apache.maven.plugins</groupId>
-                    <artifactId>maven-enforcer-plugin</artifactId>
-                    <version>3.0.0</version>
-                </plugin>
                 <!-- Self trace. -->
                 <plugin>
                     <groupId>org.codehaus.mojo</groupId>
                     <artifactId>exec-maven-plugin</artifactId>
                     <version>3.0.0</version>
-                </plugin>
-                <plugin>
-                    <groupId>org.pitest</groupId>
-                    <artifactId>pitest-maven</artifactId>
-                    <version>1.7.3</version>
                 </plugin>
                 <!--This plugin's configuration is used to store Eclipse 
                     m2e settings only. It has no influence on the Maven build itself. -->

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -294,15 +294,23 @@
             <plugin>
                 <groupId>org.jacoco</groupId>
                 <artifactId>jacoco-maven-plugin</artifactId>
+                <version>0.8.8</version>
                 <executions>
                     <execution>
+                        <id>prepare-agent</id>
                         <goals>
                             <goal>prepare-agent</goal>
                         </goals>
                     </execution>
                     <execution>
+                        <id>prepare-agent-integration</id>
+                        <goals>
+                            <goal>prepare-agent-integration</goal>
+                        </goals>
+                    </execution>
+                    <execution>
                         <id>report</id>
-                        <phase>test</phase>
+                        <phase>verify</phase>
                         <goals>
                             <goal>report</goal>
                         </goals>
@@ -319,6 +327,7 @@
                     </systemPropertyVariables>
                     <useModulePath>true</useModulePath>
                     <argLine>
+                        @{argLine}
                         --illegal-access=deny
                         <!-- Required to fix failing EqualsVerifier test:
                             TestTracedLink.testEqualsAndHashContract:22 EqualsVerifier found a problem in class org.itsallcode.openfasttrace.api.core.TracedLink.

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -14,7 +14,6 @@
         <maven.surefire.version>3.0.0-M7</maven.surefire.version>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-        <jacoco.version>0.8.8</jacoco.version>
     </properties>
     <licenses>
         <license>

--- a/product/src/test/java/org/itsallcode/openfasttrace/ITestSelfTrace.java
+++ b/product/src/test/java/org/itsallcode/openfasttrace/ITestSelfTrace.java
@@ -65,8 +65,7 @@ class ITestSelfTrace
     {
         final Path baseDir = Paths.get(new StandardDirectoryService().getCurrent()).resolve("..");
         return ImportSettings.builder() //
-                .addInputs(baseDir.resolve("doc/design.md")) //
-                .addInputs(baseDir.resolve("doc/system_requirements.md")) //
+                .addInputs(baseDir.resolve("doc/spec/")) //
                 .addInputs(findInputDirectories(baseDir)) //
                 .build();
     }

--- a/product/src/test/java/org/itsallcode/openfasttrace/mode/AbstractOftTest.java
+++ b/product/src/test/java/org/itsallcode/openfasttrace/mode/AbstractOftTest.java
@@ -20,7 +20,7 @@ abstract class AbstractOftTest
     protected final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
     protected final ByteArrayOutputStream error = new ByteArrayOutputStream();
 
-    protected void perpareOutput(final Path outputDir) throws UnsupportedEncodingException
+    protected void prepareOutput(final Path outputDir) throws UnsupportedEncodingException
     {
         this.docDir = Paths.get("../core/src/test/resources/markdown").toAbsolutePath();
         this.outputFile = outputDir.resolve("stream.txt");

--- a/product/src/test/java/org/itsallcode/openfasttrace/mode/ITestOftAsConverter.java
+++ b/product/src/test/java/org/itsallcode/openfasttrace/mode/ITestOftAsConverter.java
@@ -36,6 +36,6 @@ class ITestOftAsConverter extends AbstractOftTest
     private void assertStandardFileExportResult() throws IOException
     {
         assertOutputFileExists(true);
-        assertOutputFileContentStartsWith(SPECOBJECT_PREAMBLE + "\n  <specobjects doctype=\"dsn\">");
+        assertOutputFileContentStartsWith(SPECOBJECT_PREAMBLE + "\n  <specobjects doctype=\"");
     }
 }

--- a/product/src/test/java/org/itsallcode/openfasttrace/mode/ITestOftAsConverter.java
+++ b/product/src/test/java/org/itsallcode/openfasttrace/mode/ITestOftAsConverter.java
@@ -20,7 +20,7 @@ class ITestOftAsConverter extends AbstractOftTest
     @BeforeEach
     void beforeEach(@TempDir final Path tempDir) throws UnsupportedEncodingException
     {
-        perpareOutput(tempDir);
+        prepareOutput(tempDir);
         this.oft = Oft.create();
     }
 

--- a/product/src/test/java/org/itsallcode/openfasttrace/mode/ITestOftAsReporter.java
+++ b/product/src/test/java/org/itsallcode/openfasttrace/mode/ITestOftAsReporter.java
@@ -30,7 +30,7 @@ class ITestOftAsReporter extends AbstractOftTest
     @BeforeEach
     void beforeEach(@TempDir final Path tempDir) throws UnsupportedEncodingException
     {
-        perpareOutput(tempDir);
+        prepareOutput(tempDir);
         final ImportSettings settings = ImportSettings.builder().addInputs(this.docDir).build();
         this.oft = Oft.create();
         final List<SpecificationItem> items = this.oft.importItems(settings);

--- a/reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/OriginLinkFormatter.java
+++ b/reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/OriginLinkFormatter.java
@@ -92,16 +92,14 @@ public final class OriginLinkFormatter
 
     private static URI convertPathWithProtocolToUri(final String path)
     {
-        URI uri;
         try
         {
-            uri = URI.create(path);
+            return URI.create(path);
         }
         catch (final IllegalArgumentException e)
         {
-            uri = null;
+            return null;
         }
-        return uri;
     }
 
     private static URI convertPathWithoutProtocolToUri(final String path)

--- a/reporter/html/src/test/java/org/itsallcode/openfasttrace/report/html/view/html/TestOriginLinkFormatter.java
+++ b/reporter/html/src/test/java/org/itsallcode/openfasttrace/report/html/view/html/TestOriginLinkFormatter.java
@@ -1,7 +1,6 @@
 package org.itsallcode.openfasttrace.report.html.view.html;
 
 import static org.hamcrest.MatcherAssert.assertThat;
-
 import static org.hamcrest.core.IsEqual.equalTo;
 import static org.itsallcode.openfasttrace.report.html.view.html.OriginLinkFormatter.formatAsBlock;
 import static org.itsallcode.openfasttrace.report.html.view.html.OriginLinkFormatter.formatAsSpan;
@@ -45,6 +44,13 @@ class TestOriginLinkFormatter
                 "<a href=\"http://example.org\">http://example.org</a>:1");
     }
 
+    @Test
+    void testFormatHttpLinkWithSpecialCharacters()
+    {
+        assertPathAndLineRenderedToBlock("http://äöü.org", 1,
+                "<a href=\"http://äöü.org\">http://äöü.org</a>:1");
+    }
+
     private void assertPathAndLineRenderedToBlock(final String path, final int line,
             final String expected)
     {
@@ -76,17 +82,31 @@ class TestOriginLinkFormatter
     }
 
     @Test
-    void testFormatAbsoluteUnixPathWithSpecialCharacters()
+    void testFormatAbsoluteUnixPathWithSpecialCharactersLinux()
     {
-        OsDetector.assumeRunningOnUnix();
+        OsDetector.assumeRunningOnLinux();
         assertPathAndLineRenderedToSpan("/fo o/bär/baz", 12345678,
                 "<a href=\"file:///fo%20o/b%C3%A4r/baz\">/fo o/bär/baz</a>:12345678");
+    }
+
+    @Test
+    void testFormatAbsoluteUnixPathWithSpecialCharactersMacOs()
+    {
+        OsDetector.assumeRunningOnMacOs();
+        assertPathAndLineRenderedToSpan("/fo o/bär/baz", 12345678,
+                "<a href=\"file:///fo%20o/ba%CC%88r/baz\">/fo o/bär/baz</a>:12345678");
     }
 
     @Test
     void testFormatRelativeUnixPath()
     {
         assertPathAndLineRenderedToSpan("foo/bar/baz", 2, "foo/bar/baz:2");
+    }
+
+    @Test
+    void testFormatRelativeUnixPathWithSpecialCharacters()
+    {
+        assertPathAndLineRenderedToSpan("foo/bär/baz", 2, "foo/bär/baz:2");
     }
 
     @Test

--- a/testutil/src/main/java/org/itsallcode/openfasttrace/testutil/OsCheck.java
+++ b/testutil/src/main/java/org/itsallcode/openfasttrace/testutil/OsCheck.java
@@ -1,0 +1,53 @@
+package org.itsallcode.openfasttrace.testutil;
+
+import java.util.Locale;
+
+/**
+ * Helper class to check the operating system this Java VM runs in.
+ *
+ * please keep the notes below as a pseudo-license:
+ *
+ * http://stackoverflow.com/questions/228477/how-do-i-programmatically-determine-operating-system-in-java
+ * compare to
+ * http://svn.terracotta.org/svn/tc/dso/tags/2.6.4/code/base/common/src/com/tc/util/runtime/Os.java
+ * http://www.docjar.com/html/api/org/apache/commons/lang/SystemUtils.java.html
+ */
+public class OsCheck
+{
+    /**
+     * Types of Operating Systems
+     */
+    public enum OSType
+    {
+        WINDOWS, MACOS, LINUX, OTHER
+    }
+
+    /**
+     * Detect the operating system from the {@code os.name} System property and
+     * cache the result.
+     * 
+     * @return the operating system detected
+     */
+    public OSType getOperatingSystemType()
+    {
+        return detectOperatingSystemType();
+    }
+
+    private static OSType detectOperatingSystemType()
+    {
+        final String os = System.getProperty("os.name", "generic").toLowerCase(Locale.ENGLISH);
+        if ((os.indexOf("mac") >= 0) || (os.indexOf("darwin") >= 0))
+        {
+            return OSType.MACOS;
+        }
+        else if (os.indexOf("win") >= 0)
+        {
+            return OSType.WINDOWS;
+        }
+        else if (os.indexOf("linux") >= 0)
+        {
+            return OSType.LINUX;
+        }
+        return OSType.OTHER;
+    }
+}

--- a/testutil/src/main/java/org/itsallcode/openfasttrace/testutil/OsDetector.java
+++ b/testutil/src/main/java/org/itsallcode/openfasttrace/testutil/OsDetector.java
@@ -3,9 +3,11 @@ package org.itsallcode.openfasttrace.testutil;
 import static org.junit.jupiter.api.Assumptions.assumeFalse;
 import static org.junit.jupiter.api.Assumptions.assumeTrue;
 
+import org.itsallcode.openfasttrace.testutil.OsCheck.OSType;
+
 public class OsDetector
 {
-    private static final String OS = System.getProperty("os.name").toLowerCase();
+    private static OsCheck OS_CHECK = new OsCheck();
 
     private OsDetector()
     {
@@ -34,16 +36,16 @@ public class OsDetector
 
     private static boolean runningOnWindows()
     {
-        return OS.contains("win");
+        return OS_CHECK.getOperatingSystemType() == OSType.WINDOWS;
     }
 
     private static boolean runningOnLinux()
     {
-        return OS.contains("linux");
+        return OS_CHECK.getOperatingSystemType() == OSType.LINUX;
     }
 
     private static boolean runningOnMac()
     {
-        return OS.contains("mac");
+        return OS_CHECK.getOperatingSystemType() == OSType.MACOS;
     }
 }

--- a/testutil/src/main/java/org/itsallcode/openfasttrace/testutil/OsDetector.java
+++ b/testutil/src/main/java/org/itsallcode/openfasttrace/testutil/OsDetector.java
@@ -22,8 +22,28 @@ public class OsDetector
         assumeFalse(OsDetector::runningOnWindows, "not running on unix");
     }
 
+    public static void assumeRunningOnLinux()
+    {
+        assumeTrue(OsDetector::runningOnLinux, "not running on linux");
+    }
+
+    public static void assumeRunningOnMacOs()
+    {
+        assumeTrue(OsDetector::runningOnMac, "not running on macOS");
+    }
+
     private static boolean runningOnWindows()
     {
         return OS.contains("win");
+    }
+
+    private static boolean runningOnLinux()
+    {
+        return OS.contains("linux");
+    }
+
+    private static boolean runningOnMac()
+    {
+        return OS.contains("mac");
     }
 }

--- a/testutil/src/main/java/org/itsallcode/openfasttrace/testutil/OsDetector.java
+++ b/testutil/src/main/java/org/itsallcode/openfasttrace/testutil/OsDetector.java
@@ -7,7 +7,7 @@ import org.itsallcode.openfasttrace.testutil.OsCheck.OSType;
 
 public class OsDetector
 {
-    private static OsCheck OS_CHECK = new OsCheck();
+    private static final OsCheck OS_CHECK = new OsCheck();
 
     private OsDetector()
     {

--- a/testutil/src/test/java/org/itsallcode/openfasttrace/testutil/OsCheckTest.java
+++ b/testutil/src/test/java/org/itsallcode/openfasttrace/testutil/OsCheckTest.java
@@ -1,0 +1,72 @@
+package org.itsallcode.openfasttrace.testutil;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+import org.itsallcode.openfasttrace.testutil.OsCheck.OSType;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+class OsCheckTest {
+    
+    private static final String OS_NAME_SYSTEM_PROPERTY = "os.name";
+    private OsCheck osCheck;
+
+    @BeforeEach
+    void setup()
+    {
+        osCheck = new OsCheck();
+    }
+
+    @Test
+    void getOperatingSystemType()
+    {
+        assertThat(osCheck.getOperatingSystemType(), allOf(notNullValue(), not(equalTo(OSType.OTHER))));
+    }
+
+    @ParameterizedTest
+    @CsvSource(nullValues =
+    { "NULL" }, value = {
+            "Mac OS X, MACOS",
+            "MAC OS X, MACOS",
+            "Mac, MACOS",
+            "_Mac_, MACOS",
+            "Darwin, MACOS",
+            "Windows 10, WINDOWS",
+            "Windows 11, WINDOWS",
+            "Win, WINDOWS",
+            "_Win_, WINDOWS",
+            "linux, LINUX",
+            "LINUX DEBIAN, LINUX",
+            "unix, OTHER",
+            "NULL, OTHER"
+    })
+
+    void detectOperatingSystemType(final String osNameSystemProperty, final OSType expectedType)
+    {
+        final String orgValue = System.getProperty(OS_NAME_SYSTEM_PROPERTY);
+        try
+        {
+            setSystemProperty(osNameSystemProperty);
+            assertThat(osCheck.getOperatingSystemType(), equalTo(expectedType));
+        }
+        finally
+        {
+            setSystemProperty(orgValue);
+        }
+    }
+
+    private void setSystemProperty(final String value)
+    {
+        if (value == null)
+        {
+            System.clearProperty(OS_NAME_SYSTEM_PROPERTY);
+        }
+        else
+        {
+            System.setProperty(OS_NAME_SYSTEM_PROPERTY, value);
+        }
+    }
+}

